settings: Add a new `topics_policy` realm setting.

Vector73 opened the pull request:
Added a new realm setting, `topics_policy` and deprecated `mandatory_topics` setting as it now replaced by `topics_policy`.

Fixes part of #33549.

<!-- Describe your pull request here.-->
![Screenshot 2025-03-25 104022](https://github.com/user-attachments/assets/66a91f68-6d41-49f5-92f9-03db1158fa9f)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/models/realms.py:
```diff
@@ -155,6 +155,11 @@ class MessageEditHistoryVisibilityPolicyEnum(Enum):
 
+class RealmTopicsPolicyEnum(Enum):
+    allow_empty_topic = 2
+    disable_empty_topic = 3
```
How did you pick these numbers?

Vector73 replied:
I was thinking of reserving 1 for the `inherit` value of per-channel `topics_policy`. So, I assigned 2 and 3 to `allow_empty_topic` and `disable_empty_topic` respectively.

timabbott replied:
Yeah, that is probably reasonable. Let's talk about in the API design channel, though.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -490,6 +491,11 @@ def normalize(state: dict[str, Any]) -> None:
+                state["realm_mandatory_topics"] = (
+                    self.user_profile.realm.topics_policy
+                    == RealmTopicsPolicyEnum.disable_empty_topic.value
+                )
```
Why is this here? I think every block in this function should have a comment, but I also would expect code here to be papering over the API code itself being wrong.

timabbott commented on zerver/migrations/0699_alter_realm_topics_policy.py:
```diff
@@ -0,0 +1,17 @@
+        migrations.AlterField(
+            model_name="realm",
+            name="topics_policy",
+            field=models.PositiveSmallIntegerField(default=2),
```
It's possible we want to define this in this file and then use that, for readability. We try to avoid ever putting raw integers in these files.
```
class RealmTopicsPolicyEnum(Enum):
    allow_empty_topic = 2
```

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -123,7 +124,10 @@ def validate_message_edit_payload(
+        message.realm.topics_policy == RealmTopicsPolicyEnum.disable_empty_topic.value
+        and topic_name in ("(no topic)", "")
+    ):
         raise JsonableError(_("Topics are required in this organization."))
```
@prakhar1144 while reviewing this, I noticed we still have an explicit `"(no topic)"` check here -- should that have been updated to have that topic name get coalesced into `""` before we get here?

prakhar1144 replied:
[#backend > general chat - migration discussion @ 💬](https://chat.zulip.org/#narrow/channel/3-backend/topic/general.20chat.20-.20migration.20discussion/near/2135869)

timabbott commented on web/src/stream_popover.ts:
```diff
@@ -782,7 +787,10 @@ export async function build_move_topic_to_stream_popover(
+        if (
+            stream_data.get_stream_topics_policy(current_stream_id) ===
+            settings_config.realm_topics_policy_values.allow_empty_topic.code
+        ) {
```
This should probably get abstracted into a common function in a prep commit / PR. Maybe something like a `stream_data.can_use_general_chat(stream_id)` function that these code paths can call?

timabbott replied:
I don't think I've seen this prep commit / PR yet?

Vector73 replied:
I have made a prep commit now.

alya commented:
The strings in the screenshots are not what was specified in the issue.

alya commented:
The italics styling on "general chat" is missing.

Vector73 commented:
> The italics styling on "general chat" is missing.

Oh, I missed that. Fixed now.

alya commented:
Can we italicize it in the dropdown too?

alya commented:
Let's also change it to "Allow posting to the _general chat_ topic?"; I think that'll work better.

Vector73 commented:
> Let's also change it to "Allow posting to the _general chat_ topic?"; I think that'll work better.

Updated.



> Can we italicize it in the dropdown too?

The `<i>` tag and `font-style: italic` don't work inside `<option>` tag. So, I am unable to italicize it in the dropdown.

alya commented:
I turned off "general chat" by default, but enabled it in one channel, and sent a message there. That went fine, but when I went to look at other channels, the compose box topic field behavior is totally busted.

The topic field behavior should depend on the channel selected; it can be based on the org default when there's no channel selected. Please test carefully; you'll want to look at other places as well, like moving messages and in-line topic editing.

![Screenshot 2025-04-13 at 00 38 14@2x](https://github.com/user-attachments/assets/ec1712aa-c216-4f88-b330-9b9a6350e926)
![Screenshot 2025-04-13 at 00 36 48@2x](https://github.com/user-attachments/assets/262905b2-1bc1-4b43-9e29-6210454850bb)

Vector73 commented:
Oh, I missed that case. It happened because the placeholder text and topic input field were not updating for the streams with "No _general chat_ topic" property. Fixed it now.

timabbott commented:
https://developer.mozilla.org/en-US/docs/Learn_web_development/Extensions/Forms/Customizable_select might allow customizing the formatting, but it does seem @Vector73 is correct about the `<select>/<option>` elements. It's probably worth a discussion in #frontend as to how we want to solve that problem.

timabbott commented on zerver/views/realm.py:
```diff
@@ -91,6 +92,15 @@ def parse_message_edit_history_visibility_policy(
+    try:
+        return RealmTopicsPolicyEnum[policy_name]
+    except KeyError:
+        raise JsonableError(_("Invalid {var_name}").format(var_name="topics_policy"))
```
I think we figured out a better pattern for enums that doesn't require this type of function.

Vector73 replied:
That PR got merged. So, I have updated the code here to use the new structure now.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -111,6 +112,13 @@ def do_set_realm_property(
+            type="realm",
+            op="update",
+            property=name,
+            value=RealmTopicsPolicyEnum(value).name,
```
Hmm, should this be using `update_dict` and also sending the updated `mandatory_topics` value? I think that might be the correct thing to do rather than the `apply_events` code you have, since a naive client won't learn about the change in `mandatory_topics` without an event doing that.

Vector73 replied:
OK, changed this to `update_dict` and removed code from the `apply_event` function.

timabbott commented on api_docs/unmerged.d/ZF-d9265f:
```diff
@@ -0,0 +1,6 @@
+  Added a new realm setting, `topics_policy` defining the
+  default configuration for sending messages in empty topics.
+* [`GET /events`](/api/get-events), [`POST /register`](/api/register-queue):
+  Deprecated `mandatory_topics` field in favor of `topics_policy` realm setting.
```
You need a third entry here that you removed the `mandatory_topics` parameter from `PATCH /realm`.

Vector73 replied:
Added: 
```md
`PATCH /realm`: Removed `mandatory_topics` field as it is now replaced by `topics_policy` field.
```

timabbott commented on web/src/compose_recipient.ts:
```diff
@@ -370,6 +363,14 @@ export function update_topic_displayed_text(topic_name = "", has_topic_focus = f
+    }
+    // Otherwise, we have some adjustments to make to display:
+    // * a placeholder with the default topic name stylized
+    // * the empty string topic stylized
```
Why did this move within this function? It would be helpful for this part to be its own commit.

Vector73 replied:
This was moved to fix the following issue: https://github.com/zulip/zulip/pull/34150#issuecomment-2799843144

Previously, if we first send a message in *general chat* topic of a channel with "*General chat* allowed" topics policy and then move to another channel where *general chat* is not allowed, the topics field and placeholder were not properly reset, thus showing the topic field text in italics and making the placeholder text displayed as *general chat* instead of *Topic*.

The fix is to reset the topic field first, before executing the rest of the logic.

I have moved to this change to a separate commit.

timabbott commented on web/src/compose_tooltips.ts:
```diff
@@ -236,7 +236,10 @@ export function initialize(): void {
+                instance.setContent(parse_html(error_message));
+                // `display: flex` doesn't show the tooltip content inline when <i>general chat</i>
+                // is in the error message.
+                $(instance.popper).find(".tippy-content").css("display", "block");
```
Similarly, changing `get_disabled_send_tooltip` to HTML would be nice in its own commit.

Vector73 replied:
Moved to  separate commit.

timabbott commented on web/src/compose_validate.ts:
```diff
@@ -831,7 +835,7 @@ function validate_stream_message(scheduling_message: boolean, show_banner = true
             );
             if (is_validating_compose_box) {
-                disabled_send_tooltip_message = TOPICS_REQUIRED_ERROR_MESSAGE;
+                disabled_send_tooltip_message = TOPICS_REQUIRED_ERROR_MESSAGE.toString();
```
Put a `_html` suffix on the name of variables that are HTML, not text.

timabbott commented on web/src/stream_popover.ts:
```diff
@@ -762,12 +766,16 @@ export async function build_move_topic_to_stream_popover(
+        } else {
+            $topic_not_mandatory_placeholder.removeClass(
+                "move-topic-new-topic-placeholder-visible",
+            );
```
@prakhar1144 can your review this bit of logic? I seem to recall it being fiddly.

prakhar1144 replied:
@Vector73 can you explain a bit why is this change needed when adding this setting. Are you trying to resolve an existing bug, in that case it should be a separate commit.

Vector73 replied:
Previously, in the "Move messages" modal, switching from a channel with the `allow_empty_topic` policy to one with the `disable_empty_topic` policy did not correctly update the topic input placeholder. This change fixes that issue.

Vector73 replied:
This condition is no longer needed -- probably left over from earlier debugging.

timabbott commented on web/src/stream_ui_updates.ts:
```diff
@@ -340,6 +340,10 @@ export function enable_or_disable_permission_settings_in_edit_panel(
         }
     }
+
+    if (!settings_data.user_can_set_topics_policy() && !current_user.is_admin) {
```
`settings_data.user_can_set_topics_policy` should probably be true for admins, so that you don't need the extra conditional here? You might also want it to take the `stream_id` as a parameter so this code can consider if they're a channel admin?  I'm a bit confused by this logic.

Vector73 replied:
Oh, I created this function for testing purposes, but forgot to change it afterwards. Fixed it now.

timabbott commented on web/tests/compose_fade.test.cjs:
```diff
@@ -12,15 +12,6 @@ mock_jquery((selector) => {
-                        default:
-                            throw new Error(`Unknown arg ${arg}`);
-                    }
-                },
```
I'm not sure how this change relates to the rest of the PR.

Vector73 replied:
I removed it to pass the coverage tests, but removing only the `is(arg)` function also passes the tests.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -129,11 +129,18 @@ def validate_message_edit_payload(
+        if stream_id is None:
+            stream = Stream.objects.get(id=message.recipient.type_id)
+        else:
+            stream = Stream.objects.get(id=stream_id)
```
We should avoid adding an extra database query here; I think the bug is we're just doing this check too early.

`get_stream_by_id_in_realm` is the right function to use. But I think probably this logic should be moved to `build_message_edit_request` in a prep commit; that latter function already fetches the `Stream` object.

Vector73 replied:
I have moved this code to `build_message_edit_request`.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -1665,6 +1667,52 @@ def do_change_stream_message_retention_days(
 
 
+@transaction.atomic(durable=True)
+def do_change_stream_topics_policy(
```
For `MessageEditHistoryVisibilityPolicyEnum`, we did it as just new support in `do_set_realm_property`; I think it might be better to write this a bit more generically as a `do_set_stream_propery` in that same style? 

We don't want to end up with a million functions like this one for each setting, but I think this is the only one that's not group-based and doesn't have its own special behavior.

(This version is OK, but it seems worth improving if we're doing another round-trip of review)

Vector73 replied:
Created `do_set_stream_property` function.

timabbott commented on zerver/views/streams.py:
```diff
@@ -674,6 +700,10 @@ def add_subscriptions_backend(
+        if topics_policy is not None:
+            if not user_profile.can_set_topics_policy():
+                raise JsonableError(_("You do not have permission to change empty topics policy."))
+            stream_dict_copy["topics_policy"] = StreamTopicsPolicyEnum[topics_policy].value
```
Hmm, do we really want to disallow setting `topics_policy` to `inherit` explicitly (rather than via `none`) for users without the permission? I feel like that might be annoying. It might be reasonable to allow an explicit "inherit" during channel creation here. (Probably that rule should not apply for changing it).

Vector73 replied:
That sounds reasonable. Updated the code.

Vector73 replied:
That sounds good. Updated the code.

timabbott commented on zerver/models/streams.py:
```diff
@@ -147,6 +154,16 @@ class Stream(models.Model):
+        StreamTopicsPolicyEnum.allow_empty_topic.value: gettext_lazy(
+            "*General chat* topic allowed"
+        ),
+        StreamTopicsPolicyEnum.disable_empty_topic.value: gettext_lazy("No *general chat* topic"),
```
Let's add a comment noting that this is used for the channel events notifications when this setting is changed.

Vector73 replied:
Added the comment:
```py
    # This is used in channel events notification message when the `topics_policy`
    # is changed.
```

timabbott commented on zerver/actions/streams.py:
```diff
@@ -1665,6 +1667,52 @@ def do_change_stream_message_retention_days(
+            stream,
+            str(Realm.STREAM_EVENTS_NOTIFICATION_TOPIC_NAME),
+            _(
+                "{user_name} changed the empty topic policy from {old_topics_policy} to {new_topics_policy}."
```
Can you post a screenshot of what this looks like?

Vector73 replied:
![image](https://github.com/user-attachments/assets/e1885286-c1cc-421b-b710-fcb1b58e09c2)

Vector73 commented:
Updates:
- Fixed `user_can_set_topics_policy` function.
- Changed `op` field of `topics_policy` change event from `update` to `update_dict`.
- Created a generic `do_set_stream_property` function for changing channel fields instead of `do_change_stream_topics_policy` function.
- Moved `topics_policy` validation logic to `build_message_edit_request`.

prakhar1144 commented on web/src/stream_data.ts:
```diff
@@ -992,6 +992,14 @@ export function get_streams_for_admin(): StreamSubscription[] {
     return subs;
 }
 
+export function can_use_general_chat(stream_id: number | undefined): boolean {
```
One important naming decision was to not use the "general chat" keyword directly in function names / variable names etc. We use "empty_topic_name" / "empty_string_topic".

There's a possibility that we'll call the "general chat" something else in future. The feature was designed in such a way that we need to just change the value of a constant to do that migration. See `realm_empty_topic_display_name`.

Naming functions with "general chat" in it would result in an extra effort to rename these functions at that time. So let's stick with the generic name and use "general chat" only in user facing strings -- like settings name.

Vector73 replied:
OK, changed `can_use_general_chat` to `can_use_empty_topic`.

prakhar1144 replied:
Umm, it's still `can_use_general_chat`. I think you somehow missed to commit your changes before pushing :)

Vector73 replied:
Where? I cannot find any `can_use_general_chat` in the current changes.

prakhar1144 replied:
Oh, you renamed it in the second commit. I think it should be in the first commit itself -- no need to introduce a wrong name and then fix it in the next commit.

Vector73 replied:
Oh, forgot there was even a commit before `realm_topics_policy` one :). Fixed it now.

prakhar1144 commented on web/src/admin.ts:
```diff
@@ -32,7 +33,11 @@ const admin_settings_label = {
-    realm_mandatory_topics: $t({defaultMessage: "Require topics in channel messages"}),
+    realm_topics_policy: new Handlebars.SafeString(
+        $t_html({
+            defaultMessage: "Default <i>general chat</i> topic configuration for channels",
```
We should use `get_final_topic_display_name("")` to get the topic name here.

Vector73 replied:
Replaced the _general chat_ string with `get_final_topic_display_name("")` in setting labels, options, tooltips, and error banner. This required wrapping the strings with functions in a few places to avoid evaluation of translation strings at compile time, though.

prakhar1144 commented on web/src/settings_config.ts:
```diff
@@ -292,6 +293,37 @@ export const message_edit_history_visibility_policy_values = {
+                {defaultMessage: "{empty_topic_name} topic allowed"},
+                {
+                    empty_topic_name:
+                        empty_topic_name.charAt(0).toUpperCase() + empty_topic_name.slice(1),
```
`empty_topic_name` is translated. So uppercase rule isn't perfect here. I think we don't need to worry about capitalization here (confirm in CZO once), and just display the translated string as it is in italics (Note: the issue documents to have it in italics)

Vector73 replied:
OK, I have removed the capitalization. The `<i>` tag  doesn't work inside `<option>` tag. Any other `css` styling won't work because we can't isolate "general chat" as it will break translation. So, I am unable to italicize it in the dropdown.
https://github.com/zulip/zulip/pull/34150#issuecomment-2852109001

I will ask in #frontend how we want to handle it. Maybe we can try to solve it in a follow-up?

prakhar1144 replied:
> Maybe we can try to solve it in a follow-up?

You can discuss in CZO. Ideally, we'd be wrapping it in quotes (as [Alya said above](https://github.com/zulip/zulip/pull/34150#issuecomment-2851478968)) if we want to do it as a follow up.

Vector73 replied:
Started discussion: [#frontend > italicize "general chat" in topics_policy setting options](https://chat.zulip.org/#narrow/channel/6-frontend/topic/italicize.20.22general.20chat.22.20in.20topics_policy.20setting.20options/with/2180000)

timabbott replied:
Yeah let's wrap in quotes for this PR and just open an issue linking to this thread for that follow-up.

Vector73 replied:
OK, updated the PR to wrap "general chat" in quotes and updated the screenshots in the PR description.

prakhar1144 commented on web/src/message_edit.ts:
```diff
@@ -513,7 +513,13 @@ function handle_inline_topic_edit_change(elem: HTMLInputElement, stream_id: numb
+            $t(
+                {
+                    defaultMessage:
+                        "Sending messages to the {empty_topic_name} topic is not allowed in this channel.",
```
missing italics

Vector73 replied:
Fixed that in the last commit. It required some refactoring. Could you please take a look and let me know if it looks good to you?

Vector73 replied:
Thanks @prakhar1144  for explaining the semantics and context of _general chat_.

I have moved the changes (italicizing the "general chat" in the tooltip) to the 4th commit. Is that OK?

timabbott commented:
@prakhar1144 let me know when this is ready for me to try to integrate. I'm excited to keep this moving!

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -1784,10 +1785,13 @@ def check_message(
             and topic_name == ""
         ):
-            raise JsonableError(_("Topics are required in this organization"))
+            raise JsonableError(
+                _("Sending messages to the empty topic is not allowed in this channel")
```
We can use `get_topic_display_name()` instead of saying empty topic.

Vector73 replied:
OK, added a new exception `MessagesNotAllowedInEmptyTopicsError` to avoid duplication of the same string.

Vector73 commented on web/tests/dispatch.test.cjs:
```diff
@@ -538,9 +538,6 @@ run_test("realm settings", ({override}) => {
     assert_same(realm.realm_name, "new_realm_name");
 
-    event = event_fixtures.realm__update__mandatory_topics;
-    test_realm_boolean(event, "realm_mandatory_topics");
```
Added a check for `topics_policy` in `dispatch.test.cjs`.

Vector73 commented on zerver/actions/streams.py:
```diff
@@ -1681,6 +1683,69 @@ def do_change_stream_message_retention_days(
+        return
+
+    setattr(stream, name, value)
+    stream.save(update_fields=["topics_policy"])
```
It was discussed [here](https://github.com/zulip/zulip/pull/34150#discussion_r2074476084) that it would be better to create a generic function for updating stream properties instead of a special function, which would prevent the need for any future refactoring.

prakhar1144 replied:
Cool, so to keep it "generic" it should be:
> stream.save(update_fields=[name])

and later the `property` in `event` should be `name` too, right ?

Vector73 replied:
Yeah, right - missed that. Thanks for catching that.

Vector73 commented on zerver/models/streams.py:
```diff
@@ -155,6 +162,18 @@ class Stream(models.Model):
+    TOPICS_POLICY_DISPLAY_NAME_MAP: dict[int, Any] = {
+        StreamTopicsPolicyEnum.inherit.value: gettext_lazy("Automatic"),
+        StreamTopicsPolicyEnum.allow_empty_topic.value: gettext_lazy(
+            "*General chat* topic allowed"
```
Updated the code to use `get_topic_display_name()`. I needed to move the map to `actions/streams.py` to avoid import cycle.

prakhar1144 commented on zerver/actions/streams.py:
```diff
@@ -1681,6 +1683,69 @@ def do_change_stream_message_retention_days(
+            value=StreamTopicsPolicyEnum(value).name,
+            stream_id=stream.id,
+            name=stream.name,
+        )
```
Okay, this can be simplified to:
```py
if name == "topics_policy":
    event["value"] = StreamTopicsPolicyEnum(value).name
```

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -1412,6 +1408,15 @@ def build_message_edit_request(
+    ):
+        raise MessagesNotAllowedInEmptyTopicError(
+            get_topic_display_name("", user_profile.default_language)
+        )
```
This error will be raised when someone tries to edit the content of a message sent in "general chat"; the topic policy was changed to disabled after the message was sent.

prakhar1144 replied:
Also, this check should ideally be in `validate_message_edit_payload` -- I understand that'd result in extra database queries to determine `target_stream`, so this seems OK to me apart from the above bug. I'll let @timabbott comment on this.

There's a possibility of doing some refactoring to determine target_stream early on in `check_update_message` itself -- that'd save the extra query and the validation code will be at one place.

Vector73 replied:
> This error will be raised when someone tries to edit the content of a message sent in "general chat"; the topic policy was changed to disabled after the message was sent.

Updated to add another check to the `if` condition - `(is_topic_edited or not is_content_edited)`. Does this look good to you?

prakhar1144 replied:
If in an API call we're changing stream as well as content, this would not raise an error.

I think the check should be `(is_topic_edited or is_stream_edited)` ?

prakhar1144 replied:
Okay, works then.

prakhar1144 reviewed (commented):
Thanks for the iteration.

Looks good to me apart from that one bug.

timabbott commented on web/src/stream_data.ts:
```diff
@@ -988,6 +988,14 @@ export function get_streams_for_admin(): StreamSubscription[] {
+export function can_use_empty_topic(stream_id: number | undefined): boolean {
+    if (stream_id === undefined) {
+        return false;
+    }
```
What's the thinking behind this being `false` when there's no stream ID? I think that detail deserves a block comment explaining it. 

I have two broad questions:
- Is the commit introducing this a noop refactor, i.e., can you prove that in all situations, the result is the same as it was before the change? 
- Is `true` or `false` more accurate when we don't know what the channel is? Or could it depend on settings?

Vector73 replied:
These are the following cases where this function is used and `stream_id` can be `undefined`:
- **Drafts for the cases where the stream is not selected in the dropdown widget:** These are shown as `# > ` in the drafts UI. It was previously shown as `# > general chat` if `mandatory_topics` was `false`.
- **Composebox placeholder text for special views (e.g., Inbox, mentions, etc.):** In these cases, `Compose your message here` is used as a placeholder, same as before.
- **Topic input placeholder for special views:** `Topic` is used as a placeholder for these cases. Previously, if the `mandatory_topics` was `false`, *general chat* was used as the placeholder.

I think avoiding "_general chat_" until the channel is known makes sense, as "_general chat_" is associated with channels (for example, the user might select DM after clicking on the draft marked as `# > general chat` for destination).

> Is true or false more accurate when we don't know what the channel is? Or could it depend on settings?

I think `true` is not correct as it would mean that *general chat* will be shown even when the setting is set as "No _general chat_ allowed" at the organization-level. The alternative is to consult `realm_topics_policy` for these cases.

Also, this function may change in the follow-up when we add the new option for _general chat_ only messages. Not sure how we would like to structure that yet.

timabbott replied:
And then let's add comments, variables, or other clarifying logic to the three code paths I flagged. I'll post an example suggestion below.

timabbott commented on web/src/drafts.ts:
```diff
@@ -640,7 +639,7 @@ export function format_draft(draft: LocalStorageDraftWithId): FormattedDraft | u
         let draft_topic_display_name = draft.topic;
         let is_empty_string_topic = false;
-        if (draft.topic === "" && !realm.realm_mandatory_topics) {
+        if (draft.topic === "" && stream_data.can_use_empty_topic(draft.stream_id)) {
```
This is a specific code path where it's definitely possible for `stream_id` to be null -- basically if you saved a draft while that input was unset.

timabbott commented on web/src/compose_ui.ts:
```diff
@@ -436,7 +436,7 @@ export function compute_placeholder_text(opts: ComposePlaceholderOptions): strin
-            !realm.realm_mandatory_topics &&
+            stream_data.can_use_empty_topic(opts.stream_id) &&
             !$("input#stream_message_recipient_topic").is(":focus")
         ) {
```
This is another place where `stream_id` can definitely be `undefined`.

Vector73 replied:
This entire `if` condition is essentially a noop when the `stream_id` is undefined -- so no behavior change here.

timabbott commented on web/src/compose_recipient.ts:
```diff
@@ -339,7 +339,7 @@ export function update_topic_displayed_text(topic_name = "", has_topic_focus = f
-    if (realm.realm_mandatory_topics) {
+    if (!stream_data.can_use_empty_topic(compose_state.stream_id())) {
         return;
     }
```
This code path also can't rely on `stream_id` being defined.

Vector73 replied:
Commented on this case above.

timabbott commented on web/src/compose_actions.ts:
```diff
@@ -519,7 +518,7 @@ export let on_topic_narrow = (): void => {
-        ((compose_state.topic() || !realm.realm_mandatory_topics) &&
+        ((compose_state.topic() || stream_data.can_use_empty_topic(compose_state.stream_id())) &&
             compose_state.has_message_content()) ||
         compose_state.is_recipient_edited_manually()
```
Here we've checked that `compose_state.stream_name() === narrow_state.stream_name()`; I'm not sure whether it's still possible for both of those to be undefined and thus this refactor to be a behavior change. Worth checking.

Vector73 replied:
I have checked (_navigated across views (Inbox, combined feed, DMs, channels) to other views while keeping composebox open_), and `compose_state.stream_name()` and `narrow_state.stream_name()` can't be undefined at the same time in this code path.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -310,6 +305,11 @@ export function dispatch_normal_event(event) {
+                                if (key === "topics_policy") {
+                                    compose_recipient.update_topic_inputbox_on_topics_policy_change();
+                                    compose_recipient.update_compose_area_placeholder_text();
+                                }
```
I think this can use the `realm_settings[key]()` framework of calling registered helpers, rather than each feature adding a call here. I guess this should be documented better, since we've got a dozen violations of that model, but the intent is that one registers a function above rather than passing `noop`. Can be a follow-up PR to clean this up more fully, so let's not change this for this PR.

timabbott commented on web/src/stream_edit.ts:
```diff
@@ -109,6 +110,16 @@ export function get_display_text_for_realm_message_retention_setting(): string {
     );
 }
 
+export const get_stream_topics_policy_label = (): Handlebars.SafeString =>
```
Same comment here on an `_html` suffix.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1412,6 +1408,16 @@ def build_message_edit_request(
 
+    if (
+        target_topic_name in ("(no topic)", "")
+        and (is_topic_edited or not is_content_edited)
```
Why isn't this `and (is_topic_edited or is_stream_edited)`? While it's logically equivalent to what you have here, I think that reads more naturally as what we're trying to verify here?

Vector73 replied:
Yeah, that is more intuitive. Updated.

Vector73 commented on web/src/admin.ts:
```diff
@@ -130,6 +131,16 @@ export function build_page(): void {
             "https://zulip.readthedocs.io/en/latest/production/giphy-gif-integration.html";
     }
 
+    const realm_topics_policy_label = new Handlebars.SafeString(
```
Renamed it.

Vector73 commented on web/src/stream_popover.ts:
```diff
@@ -502,7 +502,7 @@ export async function build_move_topic_to_stream_popover(
         let is_disabled = false;
         if (
-            !stream_data.can_use_empty_topic(current_stream_id) &&
+            !stream_data.can_use_empty_topic(select_stream_id) &&
```
Yeah, this belongs to the first commit.

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -797,4 +798,27 @@ export function initialize(): void {
+                const error_message = compose_validate
+                    .get_topics_required_error_message()
+                    .toString();
+                instance.setContent(ui_util.parse_html(error_message));
```
Usually the way to do it is to have the function `get_topics_required_error_message` be `render_topics_required_error_message_html` and escape the user input while wrapping it in whatever HTML is required. (A simple way to do it is to use a tiny handlebars template for rendering that error message HTML -- and that's what I'd recommend in practice).

timabbott commented:
Oops, I accidentally passed this ID rather than #34889 when merging the latter PR, which results in GitHub irreversibly merging it.

You'll need to resubmit as a new PR, which might be convenient anyway since the comment thread on this is very long and I think we've resolved everything but the one follow-up marked here: https://github.com/zulip/zulip/pull/34150#discussion_r2116838237.
